File: android/app/src/main/java/com/iceblox/app/network/MapClient.kt
package com.iceblox.app.network

import com.iceblox.app.config.AppConfig
import com.iceblox.app.debug.DebugLog
import java.io.IOException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject

data class MapSighting(
    val latitude: Double,
    val longitude: Double,
    val confidence: Double,
    val seenAt: String,
    val type: String,
    val description: String?,
    val photoUrl: String?
)

class MapClient(private val client: OkHttpClient = OkHttpClient()) {
    suspend fun fetchSightings(lat: Double, lng: Double, radius: Double): Result<List<MapSighting>> =
        withContext(Dispatchers.IO) {
            val url = "${AppConfig.SERVER_BASE_URL}${AppConfig.MAP_SIGHTINGS_ENDPOINT}" +
                "?lat=$lat&lng=$lng&radius=$radius"

            val request = Request.Builder()
                .url(url)
                .get()
                .build()

            try {
                client.newCall(request).execute().use { response ->
                    if (!response.isSuccessful) {
                        return@withContext Result.failure(IOException("HTTP ${response.code}"))
                    }

                    val body = response.body?.string()
                        ?: return@withContext Result.failure(IOException("Empty response"))

                    val json = JSONObject(body)
                    val array = json.optJSONArray("sightings")
                        ?: return@withContext Result.success(emptyList())

                    val sightings = mutableListOf<MapSighting>()
                    for (i in 0 until array.length()) {
                        val obj = array.getJSONObject(i)
                        sightings.add(
                            MapSighting(
                                latitude = obj.getDouble("latitude"),
                                longitude = obj.getDouble("longitude"),
                                confidence = obj.getDouble("confidence"),
                                seenAt = obj.getString("seen_at"),
                                type = obj.getString("type"),
                                description = if (obj.has("description")) obj.getString("description") else null,
                                photoUrl = if (obj.has("photo_url")) obj.getString("photo_url") else null
                            )
                        )
                    }
                    Result.success(sightings)
                }
            } catch (e: IOException) {
                DebugLog.w(TAG, "fetchSightings error: ${e.message}")
                Result.failure(e)
            }
        }

    companion object {
        private const val TAG = "MapClient"
    }
}

File: ios/IceBloxApp/Networking/MapClient.swift
import Foundation

struct MapSighting: Codable {
    let latitude: Double
    let longitude: Double
    let confidence: Double
    let seenAt: String
    let type: String
    let description: String?
    let photoUrl: String?

    enum CodingKeys: String, CodingKey {
        case latitude, longitude, confidence, type, description
        case seenAt = "seen_at"
        case photoUrl = "photo_url"
    }
}

struct MapSightingsResponse: Codable {
    let status: String
    let sightings: [MapSighting]
}

final class MapClient {
    private let session = URLSession.shared

    func fetchSightings(lat: Double, lng: Double, radius: Double) async throws -> [MapSighting] {
        guard var components = URLComponents(
            url: AppConfig.serverBaseURL.appendingPathComponent(AppConfig.mapSightingsEndpoint),
            resolvingAgainstBaseURL: false
        ) else {
            throw URLError(.badURL)
        }
        components.queryItems = [
            URLQueryItem(name: "lat", value: String(lat)),
            URLQueryItem(name: "lng", value: String(lng)),
            URLQueryItem(name: "radius", value: String(radius))
        ]

        guard let url = components.url else {
            throw URLError(.badURL)
        }

        let (data, response) = try await session.data(from: url)

        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw URLError(.badServerResponse)
        }

        let decoded = try JSONDecoder().decode(MapSightingsResponse.self, from: data)
        return decoded.sightings
    }
}
